src/Storage/OWSOrphanedDataCleaner.m

```diff
@@ -9,82 +11,209 @@
+        [transaction enumerateKeysInCollection:TSThread.collection
+                                    usingBlock:^(NSString *_Nonnull key, BOOL *_Nonnull stop) {
+                                        [threadIds addObject:key];
+                                    }];
+    }];
+
+    NSMutableSet<NSString *> *orphanInteractionIds = [NSMutableSet new];
+    NSMutableSet<NSString *> *messageAttachmentIds = [NSMutableSet new];
+    [databaseConnection readWithBlock:^(YapDatabaseReadTransaction *_Nonnull transaction) {
+        [transaction enumerateKeysAndObjectsInCollection:TSMessage.collection
+                                              usingBlock:^(NSString *key, TSInteraction *interaction, BOOL *stop) {
+                                                  if (![threadIds containsObject:interaction.uniqueThreadId]) {
+                                                      [orphanInteractionIds addObject:interaction.uniqueId];
+                                                  }
 
-    [TSInteraction enumerateCollectionObjectsUsingBlock:^(TSInteraction *interaction, BOOL *stop) {
-        if ([interaction isKindOfClass:[TSMessage class]]) {
-            TSMessage *message = (TSMessage *)interaction;
-            if ([message hasAttachments]) {
-                for (NSString *attachmentId in message.attachmentIds) {
-                    [attachmentIdFilenames removeObjectForKey:attachmentId];
+                                                  if (![interaction isKindOfClass:[TSMessage class]]) {
+                                                      return;
+                                                  }
+                                                  TSMessage *message = (TSMessage *)interaction;
+                                                  if (message.attachmentIds.count > 0) {
+                                                      [messageAttachmentIds addObjectsFromArray:message.attachmentIds];
+                                                  }
+                                              }];
+    }];
+
+    DDLogDebug(@"attachmentIds: %zd", attachmentIds.count);
+    DDLogDebug(@"messageAttachmentIds: %zd", messageAttachmentIds.count);
+
+    NSMutableSet<NSString *> *orphanAttachmentIds = [attachmentIds mutableCopy];
+    [orphanAttachmentIds minusSet:messageAttachmentIds];
+    NSMutableSet<NSString *> *missingAttachmentIds = [messageAttachmentIds mutableCopy];
+    [missingAttachmentIds minusSet:attachmentIds];
+
+    DDLogDebug(@"orphan attachmentIds: %zd", orphanAttachmentIds.count);
+    DDLogDebug(@"missing attachmentIds: %zd", missingAttachmentIds.count);
+    DDLogDebug(@"orphan interactions: %zd", orphanInteractionIds.count);
+
+    // We need to avoid cleaning up new attachments and files that are still in the process of
+    // being created/written, so we don't clean up anything recent.
+    const NSTimeInterval kMinimumOrphanAge = 15 * 60.f;
+
+    if (shouldCleanup) {
```

Comment:
style nit: invert, and early return here avoids a lengthy conditional block, which improves readability and might help save us from a catastrophic mismatched brace in the future.

    if (!shouldCleanup) {
        return;
    }